thoth-export-server/src/xml/onix3_thoth.rs

```diff
@@ -0,0 +1,3368 @@
+                        .map_err(|e| e.into())
+                })
+            })?;
+            write_element_block("TitleDetail", w, |w| {
+                // 01 Cover title (serial)
+                write_element_block("TitleType", w, |w| {
+                    w.write(XmlEvent::Characters("01")).map_err(|e| e.into())
+                })?;
+                write_element_block("TitleElement", w, |w| {
+                    // 02 Collection level
+                    write_element_block("TitleElementLevel", w, |w| {
+                        w.write(XmlEvent::Characters("02")).map_err(|e| e.into())
+                    })?;
+                    write_element_block("PartNumber", w, |w| {
+                        w.write(XmlEvent::Characters(&self.issue_ordinal.to_string()))
+                            .map_err(|e| e.into())
+                    })?;
+                    write_element_block("TitleText", w, |w| {
+                        w.write(XmlEvent::Characters(&self.series.series_name))
+                            .map_err(|e| e.into())
+                    })
+                })
+            })
+        })
+    }
+}
+
+impl XmlElementBlock<Onix3Thoth> for WorkFundings {
+    fn xml_element<W: Write>(&self, w: &mut EventWriter<W>) -> ThothResult<()> {
+        write_element_block("Publisher", w, |w| {
+            // 16 Funding body
+            write_element_block("PublishingRole", w, |w| {
+                w.write(XmlEvent::Characters("16")).map_err(|e| e.into())
+            })?;
+            if let Some(ror) = &self.institution.ror {
+                write_element_block("PublisherIdentifier", w, |w| {
+                    // 40 ROR
+                    write_element_block("PublisherIDType", w, |w| {
+                        w.write(XmlEvent::Characters("40")).map_err(|e| e.into())
+                    })?;
+                    write_element_block("IDValue", w, |w| {
+                        w.write(XmlEvent::Characters(&ror.to_string()))
+                            .map_err(|e| e.into())
+                    })
+                })?;
+            }
+            if let Some(doi) = &self.institution.institution_doi {
+                write_element_block("PublisherIdentifier", w, |w| {
+                    // 32 FundRef DOI
+                    write_element_block("PublisherIDType", w, |w| {
+                        w.write(XmlEvent::Characters("32")).map_err(|e| e.into())
+                    })?;
+                    write_element_block("IDValue", w, |w| {
+                        w.write(XmlEvent::Characters(&doi.to_string()))
+                            .map_err(|e| e.into())
+                    })
+                })?;
+            }
+            write_element_block("PublisherName", w, |w| {
+                w.write(XmlEvent::Characters(&self.institution.institution_name))
+                    .map_err(|e| e.into())
+            })?;
+            let mut identifiers: HashMap<String, String> = HashMap::new();
```

Comment:
This could be done with a vector, instead – and without all the if let some

```
let fields: Vec<(&str, &Option<&str>)> = vec![
    ("programname", &self.program.as_deref()),
    ("projectname", &self.project_name.as_deref()),
    ("projectshortname", &self.project_shortname.as_deref()),
    ("grantnumber", &self.grant_number.as_deref()),
    ("jurisdiction", &self.jurisdiction.as_deref()),
];

if fields.iter().any(|(_, value)| value.is_some()) {
    write_element_block("Funding", w, |w| {
        for (typename, value) in &fields {
            if let Some(v) = *value {
                write_element_block("FundingIdentifier", w, |w| {
                    // 01 Proprietary
                    write_element_block("FundingIDType", w, |w| {
                        w.write(XmlEvent::Characters("01")).map_err(|e| e.into())
                    })?;
                    write_element_block("IDTypeName", w, |w| {
                        w.write(XmlEvent::Characters(typename)).map_err(|e| e.into())
                    })?;
                    write_element_block("IDValue", w, |w| {
                        w.write(XmlEvent::Characters(v)).map_err(|e| e.into())
                    })
                })?;
            }
        }
        Ok(())
    })?;
}
```